### PR TITLE
Site/app implementation details

### PR DESCRIPTION
**Pull request by cdegroot-adobe**
Corrected:
Tracking the implementation details for how a request is made to an Experience Service from the client is important for support, debugging and usage analysis. This PR provides a way of providing this information in a structured way.

Created a new general purpose Implementation schema that records the API name and Version.
Defined the Adobe Extension that sets the location for implementation details for Analytics, Target and Identity Service. Will be used in future with ExperienceEvent.

### CODE REVIEW

**Comment by jwen-adobe** [comment]
@trieloff @kstreeter @cmathis

**Comment by cdegroot-adobe** [comment]
@jwen-adobe can you please review this to see if it conforms to your expectations for an extension.

File: extensions/adobe/experience/implementations.schema.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
+  "$id": "https://ns.adobe.com/xdm-extensions/experience/analytics/analytics-extensions",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Adobe Experience Cloud ExperienceEvent Implementations",
+  "type": "object",
+  "meta:extensible": true,
```
Comment (trieloff) [inline extensions/adobe/experience/implementations.schema.json]: I would not make an extension itself extensible.

File: extensions/adobe/experience/implementations.schema.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  "$id": "https://ns.adobe.com/xdm-extensions/experience/analytics/analytics-extensions",
```
Comment (trieloff) [inline extensions/adobe/experience/implementations.schema.json]: This URI does not match the file name.

File: extensions/adobe/experience/implementations.schema.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Adobe Experience Cloud ExperienceEvent Implementations",
```
Comment (trieloff) [inline extensions/adobe/experience/implementations.schema.json]: Is this an implementation of an ExperienceEvent as the title indicates or rather an implementation of data collection that generates ExperienceEvents?

File: schemas/context/implementation.schema.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{
+  "meta:license": [
+    "Copyright 2017 Adobe Systems Incorporated. All rights reserved.",
+    "This work is licensed under a Creative Commons Attribution 4.0 International (CC BY 4.0) license",
+    "you may not use this file except in compliance with the License. You may obtain a copy",
+    "of the License at https://creativecommons.org/licenses/by/4.0/"
+  ],
+  "$id": "https://ns.adobe.com/xdm/context/implementation",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title": "Implementation",
```
Comment (trieloff) [inline schemas/context/implementation.schema.json]: It probably makes sense to split this into a separate PR. Reviewing the extension point and the extension in one go is a bit hard.

File: schemas/context/implementation.schema.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  "type": "object",
+  "meta:extensible": true,
+  "description": "Implementation of an API.",
```
Comment (trieloff) [inline schemas/context/implementation.schema.json]: This description is insufficient. The term "implementation" is used in a very different way in Adobe Digital Marketing from the rest of the world and "implementation of an API" does not clarify, but rather confuses the issue.

In Adobe Digital Marketing, an Implementation refers to either:
1. client-specific code that implements data collection or targeting using an online service that is running on a consumer's device
2. the SDK or tooling used to enable (1)
3. all of the above

In the rest of the world, "Implementation of an API" means a specific service that conforms to an API, typically describing the server-side, rather than the client side. 

This needs to be clarified – and verified (what are other marketing technology vendors calling their "implementations" – if they use the same term, it's great. If it's is an Adobeism, not so great). As this is most likely to become a multi-line description, it's better to create an `implementation.descrpiption.md` file.

File: schemas/context/implementation.schema.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  "definitions": {
+    "implementation": {
+      "properties": {
+        "xdm:name": {
+          "title": "Name",
+          "type": "string",
```
Comment (trieloff) [inline schemas/context/implementation.schema.json]: If there is a reasonable expectation that this value will be filtered or otherwise processed, it can make sense to use a URI to make the value truly unique. Otherwise `sCode` and `Analytics S-Code` will be treated as different values.

File: schemas/context/implementation.schema.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+          "description": "SDK or endpoint Name, e.g sCode, AppMeasurement."
```
Comment (trieloff) [inline schemas/context/implementation.schema.json]: It looks like we already know a couple of values that this property can take on. Documenting it with `meta:enum` will make this knowledge more readily accessible and reduce client implementation errors.

File: extensions/adobe/experience/implementations.schema.json (hunk 1 above)
Comment (cdegroot-adobe) [reply extensions/adobe/experience/implementations.schema.json]: Good point.

File: extensions/adobe/experience/implementations.schema.json (hunk 2 above)
Comment (cdegroot-adobe) [reply extensions/adobe/experience/implementations.schema.json]: Fixed.

File: extensions/adobe/experience/implementations.schema.json (hunk 3 above)
Comment (cdegroot-adobe) [reply extensions/adobe/experience/implementations.schema.json]: It is of data collection. I will improve the description.

File: schemas/context/implementation.schema.json (hunk 4 above)
Comment (cdegroot-adobe) [reply schemas/context/implementation.schema.json]: Usage and the used schemas are pretty interlinked.

File: schemas/context/implementation.schema.json (hunk 5 above)
Comment (cdegroot-adobe) [reply schemas/context/implementation.schema.json]: I agree, this is open to different interpretations. While "implementation" is not an adobe-ism as it is commonly used to define using a tag + library to add a feature to a web site, e.g google floodlight from google support "Migrate your Floodlight implementation to the global site tag".

The issue is that this does have another meaning for developers generally.

Some ideas I have considered:
- Implementation Details -> Seems to best fit the pattern as it is describing an implementation
- Tagging Implementation -> too web specific, needs to cover apps
- Code Implementation -> too generic.

I will propose "Implementation Details", I'm open to other recommendations.

File: schemas/context/implementation.schema.json (hunk 6 above)
Comment (cdegroot-adobe) [reply schemas/context/implementation.schema.json]: Yes, makes sense to state it is a URI

**Comment by cdegroot-adobe** [comment]
@trieloff can you please review these changes. Note I think the validators are not working right. There is an error on the nodejs version I think and an overly strict uri test.

File: schemas/context/implementationdetails.schema.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
+{
+  "meta:license": [
+    "Copyright 2017 Adobe Systems Incorporated. All rights reserved.",
```
Comment (trieloff) [inline schemas/context/implementationdetails.schema.json]: 2018

File: extensions/adobe/experience/implementations.schema.json [hunk 9]
```diff
@@ -0,0 +1,26 @@
+{
+  "$id": "https://ns.adobe.com/xdm-extensions/experience/implementations",
+  "$schema": "http://json-schema.org/draft-06/schema#",
+  "title":
+    "The Adobe Experience Cloud techncial implementations details for data collection",
```
Comment (trieloff) [inline extensions/adobe/experience/implementations.schema.json]: technical

File: schemas/context/implementationdetails.schema.json [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  "description":
+    "Details about the sdk, library or service used in an application's or web page's implementation of a service.",
+  "definitions": {
+    "implementationdetails": {
+      "properties": {
+        "xdm:name": {
+          "title": "Name",
+          "type": "string",
+          "description": "SDK or endpoint Name.",
+          "meta:enum": {
```
Comment (trieloff) [inline schemas/context/implementationdetails.schema.json]: The options in `meta:enum` here and in https://github.com/adobe/xdm/pull/153/files#diff-60dc52c2aa81b94815707651bf685e8aR9 are inconsistent.

**Comment by cdegroot-adobe** [comment]
@trieloff @kstreeter can you please review, I have rolled in all the recommendations I have at this point.  
Also note when this and the web PR are merged I will do a refresh of ExperienceEvent to include several example that represent a journey, rolling in all updates.

**Comment by trieloff** [comment]
Looks good to me now. @cdegroot-adobe: I've tightened the definition back to `uri` in b56905c. Let me know if that's good for you and I'll merge.

**Comment by cdegroot-adobe** [comment]
@kstreeter please merge
